### ISSUE TITLE
Not able to create second nullalbe int 🐛 

### ISSUE THREAD

**Issue by brian-knoll-micronetonline** [AutoFixture/AutoFixture#1459]
### Describe the Bug
When I try to create a class that has some nesting I get an error
```
ObjectCreationExceptionWithPath: AutoFixture was unable to create an instance from System.Nullable`1[System.Int32], most likely because it has no public constructor, is an abstract or non-public type.

Request path:
	UserQuery+Top
	  System.Nullable`1[System.Int32] NumberOne
	    System.Nullable`1[System.Int32]
```

### Scenario
About as minimal of a scenario as I could make.
```
void Main()
{
	Fixture fixture = new Fixture();
	fixture.Behaviors.Add(new GenerationDepthBehavior(2));
	var top = fixture.Create<Top>();
}

class Top
{
	public int? NumberOne {get;set;}
	public Bottom Bottom {get;set;}
}

class Bottom
{
	public int? NumberTwo {get;set;}
}

class GenerationDepthBehavior : ISpecimenBuilderTransformation
{
	public int Depth { get; }

	public GenerationDepthBehavior(int depth)
	{
		Depth = depth;
	}

	public ISpecimenBuilderNode Transform(ISpecimenBuilder builder)
	{
		return new RecursionGuard(builder, new OmitOnRecursionHandler(), new IsSeededRequestComparer(), Depth);
	}

	private class IsSeededRequestComparer : IEqualityComparer
	{
		bool IEqualityComparer.Equals(object x, object y)
		{
			return x is SeededRequest && y is SeededRequest;
		}

		int IEqualityComparer.GetHashCode(object obj)
		{
			return obj is SeededRequest ? 0 : EqualityComparer<object>.Default.GetHashCode(obj);
		}
	}
}
```

### Expected Behavior
Auto fixture generates the class

### Tasks

- [ ] Updated all AutoFixture and related libraries
- [ ] Confirmed in support forums if behavior is expected

### More Information
GenerationDepthBehavior is from a response in https://github.com/AutoFixture/AutoFixture/issues/1032 from @zvirja 

### Screenshots
NA

Thanks!

**Comment by GeraldLx** [comment]
I have no clue what you are trying to achieve, but the bug is not in AutoFixture but in your code.

There is no recursion in your classes Top/Bottom, so basically everything works like it should if you remove the addition of the behavior.

But lets assume your original problem is recursion, then your IsSeededRequestComparer is the problem. Basically you are saying that every SeededRequest is the same. So at the crash comes when a SeededRequest for Top is equal to a SeededRequest of Nullable<int>.

Just change
```
  public ISpecimenBuilderNode Transform(ISpecimenBuilder builder)
  {
    return new RecursionGuard(builder, new OmitOnRecursionHandler(), new IsSeededRequestComparer(), Depth);
  }
```
to
```
  public ISpecimenBuilderNode Transform(ISpecimenBuilder builder)
  {
    return new RecursionGuard(builder, new OmitOnRecursionHandler(), Depth);
  }
```
and also everything works as expected.

**Comment by brian-knoll-micronetonline** [comment]
Thank you for responding

I have a 1000 or so classes that i am trying to generate. they are deeply nested and many of them have recursive loops. For my test i want to go 1 or 2 levels into the nesting when generating the top level class. With this level of nesting generating 1 class might take 20min or more. So, i am trying to limit the depth of generation. After applying the proposed solution from #1032 i ran into this issue but they did not mention the issue back then.

I will give this new info a try thank you.

**Comment by brian-knoll-micronetonline** [comment]
So, the code in my initial post does not work for me because certain types count for an additional level in the nesting counting. For example `public int? NumberTwo {get;set;}` counts one level for the Nullable and then one level for the int. so the int returns omitspecimen and then the nullable creation blows up. I hacked my way to something that works well enough for my purposes. I would not recommend using it as is but will hopefully help someone else.
I took the initial post from #1032 and modified the currentDepth calculation so that a nullable int would not get a nest increase. and would potentially work out.
```
int currentDepth = -1;
var seededRequest = (SeededRequest)request;
var requestsForCurrentThread = this.GetMonitoredRequestsForCurrentThread();
if (requestsForCurrentThread.Count > 0)
{
    int depthIncrease = 1;
    if (IsType(seededRequest.Request)
        && (IsNullableValue(seededRequest.Request) || IsValue(seededRequest.Request)))
    {
        depthIncrease = 0;
    }

    currentDepth = requestsForCurrentThread.Max(x => x.Depth) + depthIncrease;
}
```
but when i brought it over to my actual usage i had a bunch of issues with generic types like dictionaries where they would add a whole bunch of nesting levels to the counter. For example a dictionary is made up of keyvalue pairs which is made up of the actual type parameters. a that point i gave up and slapped a trycatch on the return and called it good enough.
```
else
{
    try
    {
        return this.Builder.Create(request, context);
    }
    catch (Exception)
    {
        return new OmitSpecimen(); // Hack
    }
}
```
